3scale integration via Service annotations

zregvart opened the pull request:
WIP, first refactor to make the implementation easier

Fixes #3504

KurtStam commented:
@zregvart Can you give a little bit of background what problem you are trying so solve, and what it is you want feedback on?

zregvart commented:
@KurtStam 

For API provider-based integrations (and webhooks, perhaps) we're going to offer two options of exposing the service:
 - via k8s Service and via OpenShift Route, or
 - via k8s service and 3scale

What this PR is doing (currently) is preparing the ground for this choice of service export. Right now by simply changing the `Exposure` enum from NONE/DIRECT to SERVICE/ROUTE, to which later I'll add _3SCALE as an option also.

I was trying to test that my changes don't break anything and found out that the OpenShiftService test doesn't really test anything, so I refactored the test and added more test cases covering the Service vs Route export.

My main concern is that I don't break the existing OpenShift integration in the process. I guess the design of this is also up for discussion, though I would guess that it was agreed upon (the Exposure and the interaction with DeploymentConfig).

zregvart commented:
With this I feel we have the basic functionality done. More details are provided in the commit messages.

What this adds is: if the backend is configured to expose via 3scale (via environment variable `CONTROLLERS_EXPOSE_VIA3SCALE=true`, default is `false`) 3scale label and annotations will be added to the k8s Service and OpenShift Route will not be created.

What remains is:
 - determine the exposed URL of the integration (#3707)
 - provide the user a choice of exposure flavour (#3708)

Paging anyone at @syndesisio/backend for a review.

chirino commented:
Is here a downside to always adding the 3 scale annotations?  I would think
what needs to be configurable is if a route should be created.

On Mon, Oct 1, 2018 at 10:00 AM Zoran Regvart <notifications@github.com>
wrote:

> With this I feel we have the basic functionality done. More details are
> provided in the commit messages.
>
> What this adds is: if the backend is configured to expose via 3scale (via
> environment variable CONTROLLERS_EXPOSE_VIA3SCALE=true, default is false)
> 3scale label and annotations will be added to the k8s Service and OpenShift
> Route will not be created.
>
> What remains is:
>
>    - determine the exposed URL of the integration (#3707
>    <https://github.com/syndesisio/syndesis/issues/3707>)
>    - provide the user a choice of exposure flavour (#3708
>    <https://github.com/syndesisio/syndesis/issues/3708>)
>
> Paging anyone at @syndesisio/backend
> <https://github.com/orgs/syndesisio/teams/backend> for a review.
>
> —
> You are receiving this because you are on a team that was mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/syndesisio/syndesis/pull/3683#issuecomment-425917656>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGTV7A-WAC0lmH37NDuiPwrNOMGM-nqks5ugh_fgaJpZM4W8Joo>
> .
>
-- 
Hiram Chirino
Engineering | Red Hat, Inc.
hchirino@redhat.com | redhat.com
skype: hiramchirino | twitter: @hiramchirino

zregvart commented:
@chirino yeah that’s what the PR does, you can export via route+service or 3scale+service
